ecl/hqlcpp/hqlhtcpp.cpp

```diff
@@ -11616,97 +11616,124 @@ void HqlCppTranslator::generateSerializeFunction(BuildCtx & ctx, const char * fu
+        {
+            if (value->queryType()->getTypeCode() == type_alien)
             {
-                canSerialize = false;
-                break;
+                //MORE: Really should check if a self contained alien data type.
+                return false;
             }
-        }
+
+            OwnedHqlExpr serializedField = getSerializedForm(value->queryChild(1), diskAtom); // Could be internal, but may require serialized compare
+            OwnedHqlExpr mappedSelect = dataset.mapScalar(value,queryActiveTableSelector());
+            OwnedHqlExpr keyedCompare = dataset.mapScalar(&cur,queryActiveTableSelector());
+            info.keyFields.append(*LINK(serializedField));
+            info.keySelects.append(*createSelectExpr(LINK(mappedSelect->queryChild(0)), LINK(serializedField)));
+            info.datasetSelects.append(*LINK(value));
+            info.keyCompares.append(*LINK(keyedCompare));
+            info.filteredSorts.append(OLINK(cur));
+            info.allKeyCompares.append(*LINK(keyedCompare));
+        }
+        else if (value->isConstant())
+            info.allKeyCompares.append(OLINK(cur));
+        else
+            return false;
     }
 
+    bool aggressive = false;
+    // When projecting is done by the serialize() function this will be worth changing to true
+    // otherwise the extra cost of the project probably isn't likely to outweigh the extra copy
+    unsigned numToSerialize = aggressive ? info.filteredSorts.ordinality() : sorts.ordinality();
+
     //The following test will need to change if we serialize when nested fields are used (see above)
-    if (sorts.ordinality() >= getFlatFieldCount(record))
-        canSerialize = false;
+    if (numToSerialize >= getFlatFieldCount(record))
+        return false;
 
-    if (canSerialize)
-    {
-        if (canReuseLeft)
-        {
-            assertex(!generateCompares);
-            s.clear().append("virtual ISortKeySerializer * querySerialize").append(sideText).append("() { return &serializerLeft; }");
-            nestedctx.addQuoted(s);
-        }
-        else
-        {
-            StringBuffer memberName;
-            memberName.append("serializer").append(sideText);
+    info.keyRecord.setown(createRecord(info.keyFields));
+    info.keyDataset.setown(createDataset(no_anon, LINK(info.keyRecord)));
+    return true;
+}
 
-            BuildCtx classctx(nestedctx);
-            beginNestedClass(classctx, memberName, "ISortKeySerializer");
+void HqlCppTranslator::generateSerializeKey(BuildCtx & nestedctx, node_operator side, SerializeKeyInfo & keyInfo, const DatasetReference & dataset, bool generateCompares)
+{
+    //check if there are any ifblocks, and if so don't allow it.  Even more accurate would be no join fields used in ifblocks
+    //IHqlExpression * record = dataset.queryDataset()->queryRecord();
+    const char * sideText = (side == no_none) ? "" : (side == no_left) ? "Left" : "Right";
+    StringBuffer s;
```

Comment:
The use of clear() is there so that all calls are consistent.
It could be closer to where it is needed, I'm not sure it is worth changing, but could do.